hub-api/src/main/java/ch/prevo/open/encrypted/services/DataEncryptionService.java

```diff
@@ -0,0 +1,178 @@
+ * Public License, v. 2.0 are satisfied: GNU General Public License, version 3
+ * with the GNU Classpath Exception which is
+ * available at https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-3.0 WITH Classpath-exception-2.0
+ *
+ * Contributors:
+ *     Prevo-System AG - initial API and implementation
+ ******************************************************************************/
+package ch.prevo.open.encrypted.services;
+
+import ch.prevo.open.encrypted.model.EncryptedData;
+import ch.prevo.open.encrypted.model.SignedEncryptionContainer;
+import ch.prevo.open.encrypted.model.SymmetricKeyBundle;
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import javax.crypto.Cipher;
+import javax.crypto.KeyGenerator;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.IvParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.security.GeneralSecurityException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.security.SecureRandom;
+import java.security.Signature;
+import java.security.SignatureException;
+import java.util.Base64;
+
+/**
+ * Uses AES/RSA to encrypt data using a public key, and decryption using a private key.
+ * In addition the encrypted data is signed with SHA256withRSA before it is encrypted to provide a signature that will be
+ * validated during decryption of the data.
+ *
+ */
+public class DataEncryptionService {
+
+    private static final String SYMMETRIC_TRANSFORMATION_ALGORITHM = "AES";
+    private static final String SYMMETRIC_TRANSFORMATION = "AES/CBC/PKCS5Padding";
+    public static final String ASYMMETRIC_TRANSFORMATION_ALGORITHM = "RSA";
+    private static final String ASYMMETRIC_TRANSFORMATION = "RSA/ECB/OAEPWithSHA-1AndMGF1Padding";
+    private static final int SYMMETRIC_KEY_SIZE = 256;
+    private static final int SYMMETRIC_BLOCK_SIZE_BYTES = 16;
+    private static final String SIGNING_ALGORITHM = "SHA256withRSA";
+    private final ObjectMapper objectMapper = new ObjectMapper();
+
+
+    public <T> EncryptedData encryptAndSign(T data, PublicKey publicKey, PrivateKey privateKey) {
+        try {
+            SecretKey symmetricKey = createSymmetricKey();
+            IvParameterSpec iv = createIV();
+
+            String signature = sign(toByteArray(data), privateKey);
+            SignedEncryptionContainer<T> signedEncryptionContainer = new SignedEncryptionContainer<>(data, signature);
+            String encryptedDataBase64 = toBase64(encryptSymmetrically(signedEncryptionContainer, symmetricKey, iv));
+
+            SymmetricKeyBundle symmetricKeyBundle = new SymmetricKeyBundle(toBase64(symmetricKey.getEncoded()), toBase64(iv.getIV()));
```

Comment:
There is way too many toBase64-calls in my opinion.

Reply:
I was able to remove all base64 conversion since you were absolutely right that jackson converts it to base64 automatically. - makes code much more readable 👍